Sources/Basics/TemporaryFile.swift

```diff
@@ -0,0 +1,111 @@
+///            variables are set, dir will be set to `/tmp/`.
+///     - prefix: The prefix to the temporary file name.
+///     - body: A closure to execute that receives the absolute path of the directory as an argument.
+///           If `body` has a return value, that value is also used as the
+///           return value for the `withTemporaryDirectory` function.
+///           The cleanup block should be called when the temporary directory is no longer needed.
+///
+/// - Throws: `MakeDirectoryError` and rethrows all errors from `body`.
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+public func withTemporaryDirectory<Result>(
+    dir: AbsolutePath? = nil,
+    prefix: String = "TemporaryDirectory",
+    _ body: (AbsolutePath, @escaping (AbsolutePath) -> Void) async throws -> Result
+) async throws -> Result {
+    let temporaryDirectory = try createTemporaryDirectory(dir: dir, prefix: prefix)
+    return try await body(temporaryDirectory) { path in
+        _ = try? FileManager.default.removeItem(atPath: path.pathString)
+    }
+}
+
+/// Creates a temporary directory and evaluates a closure with the directory path as an argument.
+/// The temporary directory will live on disk while the closure is evaluated and will be deleted afterwards.
+///
+/// This function is basically a wrapper over posix's mkdtemp() function.
+///
+/// - Parameters:
+///     - dir: If specified the temporary directory will be created in this directory otherwise environment
+///            variables TMPDIR, TEMP and TMP will be checked for a value (in that order). If none of the env
+///            variables are set, dir will be set to `/tmp/`.
+///     - prefix: The prefix to the temporary file name.
+///     - removeTreeOnDeinit: If enabled try to delete the whole directory tree otherwise remove only if its empty.
+///     - body: A closure to execute that receives the absolute path of the directory as an argument.
+///             If `body` has a return value, that value is also used as the
+///             return value for the `withTemporaryDirectory` function.
+///
+/// - Throws: `MakeDirectoryError` and rethrows all errors from `body`.
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+public func withTemporaryDirectory<Result>(
+    dir: AbsolutePath? = nil,
+    prefix: String = "TemporaryDirectory",
+    removeTreeOnDeinit: Bool = false ,
+    _ body: (AbsolutePath) async throws -> Result
+) async throws -> Result {
+    try await withTemporaryDirectory(dir: dir, prefix: prefix) { path, cleanup in
+        defer { if removeTreeOnDeinit { cleanup(path) } }
+        return try await body(path)
+    }
+}
+
+private func createTemporaryDirectory(dir: AbsolutePath?, prefix: String) throws -> AbsolutePath {
+    // Construct path to the temporary directory.
+    let templatePath = try AbsolutePath(prefix + ".XXXXXX", relativeTo: determineTempDirectory(dir))
```

Comment:
Where is `determineTempDirectory`?

Reply:
https://github.com/apple/swift-tools-support-core/blob/503e25f60303cf6c483828512ed66be2d87e4cc7/Sources/TSCBasic/TemporaryFile.swift#L55-L61

Reply:
I think we need to add a new version of that that take `FileSystem` instead of harding to `localFileSystem`

Reply:
Done it!